google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/functions/ResetWindowsPassword.java

```diff
@@ -0,0 +1,229 @@
+      this.api = api;
+      this.crypto = crypto;
+      this.operationDone = operationDone;
+   }
+
+   @Override
+   public String apply(Map<String, ?> params) {
+      String zone = (String)params.get("zone");
+       final AtomicReference<Instance> instance = (AtomicReference<Instance>)params.get("instance");
+       userName = (String)params.get("userName");
+       email = (String)params.get("email");
+
+      // TODO Check whether VM is up
+      try {
+         // Generate the public/private key pair for encryption and decryption.
+         // TODO do we need to explicitly set 2048 bits? Presumably "RSA" is implicit
+         KeyPair keys = crypto.rsaKeyPairGenerator().genKeyPair();
+
+         // Update instance's metadata with new "windows-keys" entry, and wait for operation to 
+         // complete.
+         logger.debug("Generating windows key for instance %s, by updating metadata", instance.get().name());
+         final InstanceApi instanceApi = api.instancesInZone(zone);
+         Metadata metadata = instance.get().metadata();
+         // If disableHtmlEscaping is not there, == will be escaped from modulus value
+          metadata.put("windows-keys", new GsonBuilder().disableHtmlEscaping().create().toJson(extractKeyMetadata(keys)));
+
+         AtomicReference<Operation> operation = Atomics.newReference(instanceApi.setMetadata(instance.get().name(), metadata));
+         operationDone.apply(operation);
+
+         if (operation.get().httpErrorStatusCode() != null) {
+            logger.warn("Generating windows key for %s failed. Http Error Code: %d HttpError: %s",
+                  operation.get().targetId(), operation.get().httpErrorStatusCode(),
+                  operation.get().httpErrorMessage());
+         }
+
+         TimeLimiter limiter = new SimpleTimeLimiter();
+         String encryptedPassword = null;
+         try {
+            encryptedPassword = limiter.callWithTimeout(new Callable<String>() {
+               public String call() {
+                  String serialPortContents = "";
+                  Map<String, String> passwordDict = ImmutableMap.of();
+                  while (!serialPortContents.startsWith("{\"ready\":true") || passwordDict.get("encryptedPassword") == null) {
+                     try {
+                        Thread.sleep(30 * 1000);
+                     } catch (InterruptedException e) {
+                        Throwables.propagate(e);
+                     }
+                     serialPortContents = instanceApi.getSerialPortOutput(instance.get().name(), 4).contents();
+                     String[] contentEntries = serialPortContents.split("\n");
+                     passwordDict = new Gson().fromJson(contentEntries[contentEntries.length - 1], Map.class);
+                  }
+                  return passwordDict.get("encryptedPassword");
+               }
+            }, 5, TimeUnit.MINUTES, false); // Notice that timeoutDuration should be less than EXPIRE_DURATION
```

Comment:
Instead of manually doing this and calling Thread.sleep, use the jclouds `Predicates2.retry`.